packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -48,11 +48,40 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
   // Initialized to true as a new layer has never called [addToScene].
   bool _needsAddToScene = true;
 
+  /// Whether the parent layer is required to call this layer's [addToScene]
+  /// method to render it.
+  ///
+  /// If false, the parent layer may pass this layer's engine layer to
+  /// [SceneBuilder.addRetained]. If true, [SceneBuilder.addRetained] may not
+  /// be used.
+  bool get debugNeedsAddToScene {
+    bool result;
+    assert(() {
+      result = _needsAddToScene;
+      return true;
+    }());
+    return result;
+  }
+
   /// Mark that this layer has changed and [addToScene] needs to be called.
   @protected
   @visibleForTesting
   void markNeedsAddToScene() {
+    assert(
+      !alwaysNeedsAddToScene,
+      '$runtimeType called markNeedsAddToScene.\n'
```

Comment:
this line itself doesn't describe a problem. Consider something like "`$runtimeType with alwaysNeedsAddToScene set called markNeedsAddToScene.`" maybe?

Reply:
Done.